### PR TITLE
Fix line of sight check computation for theta* planner

### LINKED ISSUE #5863: [nav2_theta_star_planner] Theta* paths intersects lethal obstacles.
## Bug report

The `nav2_theta_star_planner/ThetaStarPlanner` produces paths that intersect lethal obstacles.

### Required Info

- Operating System: Ubuntu 22.04
- ROS2 Distro: Humble
- NAV2 Version: ros-humble-navigation2 1.1.20-1jammy.20251228.132604

### Steps to reproduce issue

1. Launch RViz2 and publish your static tf tree.
2. Use the provided `nav2_params.yaml` file to launch your `planner_server`, `bt_navigator` and `lifecycle_manager`.
3. Publish a map on the `/map` topic and use RViz2 to set a goal 2D pose.

### Expected behavior

The planner finds the cheapest path between the starting pose and the goal pose while avoiding obstacles.

### Actual behavior

The path intersects lethal obstacles.

### Configuration

`nav2_params.yaml`:

```
global_costmap:
  global_costmap:
    ros__parameters:
      global_frame: map
      robot_base_frame: base_link
      robot_radius: 0.5
      plugins: ["static_layer", "inflation_layer"]
      trinary_costmap: false
      lethal_cost_threshold: 99

      static_layer:
        plugin: "nav2_costmap_2d::StaticLayer"
        map_topic: /map
        subscribe_to_updates: false
        enabled: true
      
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        enabled: True
        cost_scaling_factor: 20.0
        inflation_radius: 1.0

planner_server:
  ros__parameters:
    use_sim_time: false
    expected_planner_frequency: 5.0
    costmap_update_timeout: 1.0
    allow_partial_planning: false
    planner_plugins: ["GridBased"]
    
    GridBased:
      plugin: "nav2_theta_star_planner/ThetaStarPlanner"
      how_many_corners: 8
      w_euc_cost: 1.0
      w_traversal_cost: 2.0

bt_navigator:
  ros__parameters:
    use_sim_time: false
    global_frame: map
    robot_base_frame: base_link
    odom_topic: odom
```

### Test Cases

- **Test 1: Lower Res Map with No Inflation Layer**
  - Size: 25 x 25 meters
  - Resolution: 1 m/px
  - Inflation Layer: None
  - Start: `x = -5.0 m`, `y = +11.0 m`
  - Goal: `x = +11.0 m`, `y = -8.0 m`

Image

```
[bt_navigator-5] [INFO] [1768318176.880994941] [bt_navigator]: Begin navigating from current location (-5.00, 11.00) to (11.00, -8.00)
[bt_navigator-5] [INFO] [1768318176.911177855] [bt_navigator]: Goal succeeded
```

**Result**: planner succeeds but path intersect lethal obstacles.

---

- **Test 2: Higher Res Map with No Inflation Layer**
  - Size: 125 x 125 meters
  - Resolution: 0.2 m/px
  - Inflation Layer: None
  - Start: `x = -5.0 m`, `y = +11.0 m`
  - Goal: `x = +11.0 m`, `y = -8.0 m`

Image

```
[bt_navigator-5] [INFO] [1768318176.880994941] [bt_navigator]: Begin navigating from current location (-5.00, 11.00) to (11.00, -8.00)
[bt_navigator-5] [INFO] [1768318176.911177855] [bt_navigator]: Goal succeeded
```

**Result**: planner succeeds but path intersect lethal obstacles.

---

- **Test 3: Higher Res Map with Inflation Layer**
  - Size: 125 x 125 meters
  - Resolution: 0.2 m/px
  - Inflation Layer: Yes (see **Configuration** section)
  - Start: `x = -5.0 m`, `y = +11.0 m`
  - Goal: `x = +11.0 m`, `y = -8.0 m`

Image

```
[bt_navigator-5] [INFO] [1768318176.880994941] [bt_navigator]: Begin navigating from current location (-5.00, 11.00) to (11.00, -8.00)
[bt_navigator-5] [INFO] [1768318176.911177855] [bt_navigator]: Goal succeeded
```

**Result**: planner succeeds but path intersect lethal regions of inflation layer.

---

- **Test 4: Narrower Map with No Inflation Layer**
  - Size: 25 x 25 meters
  - Resolution: 1.0 m/px
  - Inflation Layer: None
  - Start: `x = -5.0 m`, `y = +11.0 m`
  - Goal: `x = +11.0 m`, `y = -5.0 m`

Image

```
[bt_navigator-5] [INFO] [1768317954.959984951] [bt_navigator]: Begin navigating from current location (-5.00, 11.00) to (11.00, -5.00)
[planner_server-4] [ERROR] [1768317954.960576225] [planner_server]: Could not generate path between the given poses
[planner_server-4] [WARN] [1768317954.960621343] [planner_server]: Planning algorithm GridBased failed to generate a valid path to (10.99, -5.06)
[planner_server-4] [WARN] [1768317954.960635590] [planner_server]: [compute_path_to_pose] [ActionServer] Aborting handle.
[bt_navigator-5] [WARN] [1768317954.990128094] [bt_navigator]: [navigate_to_pose] [ActionServer] Aborting handle.
[bt_navigator-5] [ERROR] [1768317954.990193396] [bt_navigator]: Goal failed
```

**Result**: planner fails to find viable path.

### Additional information

- For all tests, the `bt_navigator` uses custom behavior trees to only include the `planner_server` in the pipeline. Robot control is ignored.

### PR DESCRIPTION
**Pull request by astrorick**
## Basic Info

| Info | |
| ------ | ----------- |
| Ticket(s) this addresses   | #5863 |
| Primary OS tested on | Ubuntu 22.04 |
| Robotic platform tested on | N/A |
| Does this PR contain AI generated software? | No |
| Was this PR description generated by AI software? | No |

---

## Description of contribution in a few bullet points

* Fixed how the `losCheck` function performs line of sight checks in `nav2_theta_star_planner/src/theta_star.cpp`.
* Slightly improved overall readability of the `losCheck` function.
* No changes to `nav2_theta_star_planner/src/theta_star_planner.cpp`.
* No changes to header files.

## Description of documentation updates required from your changes

* None

## Description of how this change was tested

* Built `navigation2` from source code (`humble_main` branch), including package `nav2_theta_star_planner`, using `colcon build`.
* Tested package `nav2_theta_star_planner` on Ubuntu 22.04 + ROS2 Humble.
  * `colcon test --packages-select nav2_theta_star_planner`:
    ```
    Starting >>> nav2_theta_star_planner
    Finished <<< nav2_theta_star_planner [0.26s]          
    
    Summary: 1 package finished [0.41s]
    ```
  * `colcon test-result --verbose`:
    ```
    Summary: 4 tests, 0 errors, 0 failures, 0 skipped
    ```

---

## Future work that may be required in bullet points

* Maybe consider validating unit tests to check whether they cover all appropriate scenarios.

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

### CODE REVIEW

File: nav2_theta_star_planner/src/theta_star.cpp [hunk 1]
```diff
@@ -181,63 +181,67 @@ void ThetaStar::backtrace(std::vector<coordsW> & raw_points, const tree_node * c
...
-  cx = x0;
-  cy = y0;
-
+  if (!isSafe(cx, cy)) {
+    return false;
+  }
+  
   if (dx >= dy) {
-    while (cx != x1) {
+    for (int i = 0; i <= dx; i++) {
```
Comment (SteveMacenski) [inline nav2_theta_star_planner/src/theta_star.cpp]: We increment cx/cy in the loop, i this is `<=` wouldn't this go past the last cell?

File: nav2_theta_star_planner/src/theta_star.cpp [hunk 2]
```diff
@@ -181,63 +181,67 @@ void ThetaStar::backtrace(std::vector<coordsW> & raw_points, const tree_node * c
...
       f += dy;
-      if (f >= dx) {
-        if (!isSafe(cx + u_x, cy + u_y, sl_cost)) {
+      if (f > 0) {        
+        if (!isSafe(cx + sx, cy)) {
```
Comment (SteveMacenski) [inline nav2_theta_star_planner/src/theta_star.cpp]: You seem to use anohter API for isSafe in many places, which affects the `sl_cost` accumulation to return; I think this is a bug here and in other locations

Comment (SteveMacenski) [reply nav2_theta_star_planner/src/theta_star.cpp]: also poses a questoin that checking on the corners would cause "double counting" of costs in some situations which would impact path planning quality/correctness

**Review by SteveMacenski — COMMENTED** [review]
Please checkout CI/linting. I can review once those pass :smile: 

I did a quick check and found some bugs, which also concerned me a bit since this bit of code is admittedly not easy for me to review and reason about fully (basically: I'm concerned that I wouldn't find all the errors)

File: nav2_theta_star_planner/src/theta_star.cpp (hunk 1 above)
Comment (astrorick) [reply nav2_theta_star_planner/src/theta_star.cpp]: Since the marching cell `(cx, cy)` only moves either horizontally OR vertically during LoS checks now, I thought that this condition was necessary to guarantee the inclusion of the last cell.

However, I tried switching from `<=` to `<` in both conditions and ran a few more tests. I included a bunch of prints in the code to see which cell was being checked at the end of the path. The change does not seem to have had any effect on final result (planner behaves and tests are passing) so maybe I was mistaken.

Will change this in the next commit.

File: nav2_theta_star_planner/src/theta_star.cpp (hunk 2 above)
Comment (astrorick) [reply nav2_theta_star_planner/src/theta_star.cpp]: Yes, I used a different API for the `isSafe()` function during explicit diagonal cell checking (required to solve corner cutting) in both `dx >= dy` and `dx < dy` cases.

If I interpreted the code correctly, I believe these additional checks should not contribute towards the accumulation of traversal cost, since we're not actually moving through that space but we're only checking lateral cells to avoid hitting corners.

If we do also accumulate cost there, we risk overestimating the cost of moving diagonally w.r.t. the octile distance (A*-like).

**Comment by astrorick** [comment]
Thank you so much for your feedback, I'll try fixing the CI/linting errors ASAP.

File: nav2_theta_star_planner/src/theta_star.cpp [hunk 3]
```diff
@@ -187,54 +187,60 @@ bool ThetaStar::losCheck(
...
+  int dy = abs(y1 - y0), sy = (y0 < y1) ? 1 : -1;
+  int cx = x0, cy = y0, f = (dx >= dy) ? -dx / 2 : -dy / 2;
 
-  int u_x = (sx - 1) / 2;
-  int u_y = (sy - 1) / 2;
-  cx = x0;
-  cy = y0;
+  if (!isSafe(cx, cy)) {
+    return false;
+  }
```
Comment (SteveMacenski) [inline nav2_theta_star_planner/src/theta_star.cpp]: Is the start point redundant if the start from the plugin is checked to be valid (I believe it is?) and then all other end points are checked -- so we know any start point should be valid.

i think we should be checking end points instead

File: nav2_theta_star_planner/src/theta_star.cpp [hunk 4]
```diff
@@ -187,54 +187,60 @@ bool ThetaStar::losCheck(
...
+        cx += sx;
         cy += sy;
         f -= dx;
+      } else {
+        cx += sx;
       }
-      if (f != 0 && !isSafe(cx + u_x, cy + u_y, sl_cost)) {
-        return false;
-      }
-      if (dy == 0 && !isSafe(cx + u_x, cy, sl_cost) && !isSafe(cx + u_x, cy - 1, sl_cost)) {
```
Comment (SteveMacenski) [inline nav2_theta_star_planner/src/theta_star.cpp:209]: We have special checks here for axially aligned (all X or all Y) segments. I think if we're checking diagonals to be conservative, doesn't this also meet that same line of thinking?

File: nav2_theta_star_planner/src/theta_star.cpp [hunk 5]
```diff
@@ -187,54 +187,60 @@ bool ThetaStar::losCheck(
...
+    return false;
+  }
 
   if (dx >= dy) {
-    while (cx != x1) {
+    for (int i = 0; i < dx; i++) {
       f += dy;
-      if (f >= dx) {
-        if (!isSafe(cx + u_x, cy + u_y, sl_cost)) {
+      if (f > 0) {
```
Comment (SteveMacenski) [inline nav2_theta_star_planner/src/theta_star.cpp]: For both instances, should this be `f >= 0`?

**Review by SteveMacenski — COMMENTED** [review]
Sorry for the delay in reviews, I was traveling last week and I needed the big screen and some quiet to really think about this

**Comment by SteveMacenski** [comment]
@astrorick any updates?

**Comment by SteveMacenski** [comment]
@astrorick any updates? I'd love to get this in!

**Comment by astrorick** [comment]
Hey @SteveMacenski, really sorry I was unable to reach out sooner. I've been busy with urgent activities related to my PhD lately, and I could not follow up on our discussion.

I'll do my best to get back to you as soon as possible, I too would love to see this integrated!

**Comment by astrorick** [comment]
Hey @SteveMacenski, just wanted to let you know I was finally able to tackle this task! I just pushed a slightly modified version that is more on the lines of the original version but with (hopefully) improved clarity.

The feedback you provided on the previous iteration was invaluable, and allowed me to finally grasp what was going on under the hood in the original version. Basically, the `isSafe` check was being called _between_ the updates on the values of `cx` and `cy`, effectively only affecting _one_ of the intermediate lateral cells. Depending on which side the obstacle is on, it may or may not catch it, and this was causing the issue.

The updated version fixes this by explicitly detecting diagonal steps and checking both lateral cells. The distance from the ideal line is now tracked by the error `e`, which replaces `f` but serves the same purpose.

This is the new `losCheck` function, with additional comments highlighting key parts:

```
bool ThetaStar::losCheck(const int & x0, const int & y0, const int & x1, const int & y1, double & sl_cost) const {
  sl_cost = 0;

  // this part has only been cleared up a bit
  int dx = abs(x1 - x0), sx = (x0 < x1) ? 1 : -1;
  int dy = abs(y1 - y0), sy = (y0 < y1) ? 1 : -1;
  int cx = x0, cy = y0, e = dx - dy;

  while (cx != x1 || cy != y1) {      // <- reverted back to while loop for clarity
    if (!isSafe(cx, cy, sl_cost)) {   // <- check safety for the cell we are currently on
      return false;
    }
    int e2 = 2 * e;              // <- this avoids floating point operations
    if (e2 > -dy && e2 < dx) {   // <- this only triggers for diagonal movements
      if (!isSafe(cx + sx, cy) || !isSafe(cx, cy + sy)) {
        return false;
      }
      cx += sx;
      cy += sy;
      e += dx - dy;
    } else if (e2 > -dy) {   // <- horizontal movement
      cx += sx;
      e -= dy;
    } else {                 // <- vertical movement
      cy += sy;
      e += dx;
    }
  }

  // when the loop breaks (target reached), check the final cell
  if (!isSafe(cx, cy, sl_cost)) {
    return false;
  }

  return true;
}
```

It is also worth mentioning that the additional checks performed on the lateral cells (in the case of diagonal movement) should not contribute to increase the value of `sl_cost`. For this reason, the alternative API for `isSafe` was used there.

Let me know if this look good to you!

**Comment by astrorick** [comment]
Hi @SteveMacenski, I was wondering if you had the chance to take a peek at last week's update.

I also noticed that CircleCI failed on a few of the tests last time. However, the failures seem to be related to `nav2_behavior_tree` and `opennav_docking`, while the `nav2_theta_star_planner` package that I worked on was not affected. I wanted to ask if it could be possible to re-run CI to rule out possible transient issues.

Thanks and happy Easter!

File: nav2_theta_star_planner/src/theta_star.cpp [hunk 6]
```diff
@@ -187,54 +187,35 @@ bool ThetaStar::losCheck(
...
-      }
-      if (f != 0 && !isSafe(cx + u_x, cy + u_y, sl_cost)) {
-        return false;
-      }
-      if (dy == 0 && !isSafe(cx + u_x, cy, sl_cost) && !isSafe(cx + u_x, cy - 1, sl_cost)) {
-        return false;
-      }
-      cx += sx;
+  int dx = abs(x1 - x0), sx = (x0 < x1) ? 1 : -1;
+  int dy = abs(y1 - y0), sy = (y0 < y1) ? 1 : -1;
```
Comment (SteveMacenski) [inline nav2_theta_star_planner/src/theta_star.cpp]: You flipped the condition on x/y0 and x/y1, do you realize that?

**Comment by SteveMacenski** [comment]
I've been really far behind on reviews sorry about that

File: nav2_theta_star_planner/src/theta_star.cpp [hunk 7]
```diff
@@ -187,54 +187,35 @@ bool ThetaStar::losCheck(
...
-          return false;
-        }
-        cx += sx;
-        f -= dy;
-      }
-      if (f != 0 && !isSafe(cx + u_x, cy + u_y, sl_cost)) {
-        return false;
-      }
-      if (dx == 0 && !isSafe(cx, cy + u_y, sl_cost) && !isSafe(cx - 1, cy + u_y, sl_cost)) {
+    int e2 = 2 * e;
```
Comment (SteveMacenski) [inline nav2_theta_star_planner/src/theta_star.cpp:193]: Also this accounts for the fact that 'diagonals' are not just 45 deg or something, but at any angle due to the nature of visibility checks in Theta*?

**Comment by SteveMacenski** [comment]
From an AI review agent that I made that seems to be right, though I reserve my final judgement (and why I didn't mention in the review) until I can fully review it myself - hopefully tomorrow

In resetParent, losCheck is called between two already-settled nodes. The new implementation calls isSafe(cx, cy, sl_cost) on both (x0, y0) and (x1, y1), accumulating their traversal cost into sl_cost. That cost is then added on top of maybe_par->g, which already includes those cells' traversal costs from when they were originally settled. This inflates the g_cost of LOS-shortcutted edges and biases the planner against taking them through high-cost areas.

Fix: Don't accumulate sl_cost for the start and end cells — only for the intermediate cells traversed by the line.

Steve's Note: I think we should keep the end, but not the start cell? The next iteration expanding from the end cell would check that as part of its start. This is how it works before this PR. I agree there's double counting with the final check

```
  if (!isSafe(cx, cy, sl_cost)) {
    return false;
  }
```

---

The diagonal branch fires when e2 > -dy && e2 < dx (strict). When e2 == dx exactly, the code falls into the x-only branch with no check of (cx+sx, cy) and (cx, cy+sy). For a line precisely at that Bresenham threshold, a corner that the robot would need to squeeze through is not validated. The y-step is deferred to the next iteration, but the flanking obstacle check that should accompany it never happens.

Fix: Change e2 < dx to e2 <= dx (or equivalently e2 > -dy && e2 < dx → e2 >= -dy && e2 < dx for the other bound), consistent with the standard Bresenham tie-breaking convention that triggers both steps simultaneously.

File: nav2_theta_star_planner/src/theta_star.cpp [hunk 8]
```diff
@@ -187,54 +187,35 @@ bool ThetaStar::losCheck(
...
-        if (!isSafe(cx + u_x, cy + u_y, sl_cost)) {
-          return false;
-        }
-        cx += sx;
-        f -= dy;
-      }
-      if (f != 0 && !isSafe(cx + u_x, cy + u_y, sl_cost)) {
-        return false;
-      }
-      if (dx == 0 && !isSafe(cx, cy + u_y, sl_cost) && !isSafe(cx - 1, cy + u_y, sl_cost)) {
```
Comment (SteveMacenski) [inline nav2_theta_star_planner/src/theta_star.cpp:227]: Then perfectly vertical or horizontal, we check some neighbors too. Should we do that here?

File: nav2_theta_star_planner/src/theta_star.cpp (hunk 7 above)
Comment (astrorick) [reply nav2_theta_star_planner/src/theta_star.cpp:193]: Yes, this should cover all cases that are not perfectly horizontal or vertical (since we are sure those won't cut corners anyway).

The branch that fires for _local_ diagonal movement, which explicitly checks for lateral obstacles, refers to a single step of the marching cell. By performing these extra checks every time the marching cell moves diagonally, we make sure no corners are cut by the _final_ line, which can be any angle.

File: nav2_theta_star_planner/src/theta_star.cpp (hunk 4 above)
Comment (astrorick) [reply nav2_theta_star_planner/src/theta_star.cpp:209]: I think these checks were placed here to handle perfectly horizontal or vertical edge cases, which were include in the `(dx >= dy)` and `(dx < dy)` branches in the old version of this function.

The new version explicitly includes branches for those two cases, which we know don't need extra checking for lateral obstacles during movement.

I'm not sure I was able to answer appropriately to this question, please feel free to iterate on it if necessary.

File: nav2_theta_star_planner/src/theta_star.cpp (hunk 8 above)
Comment (astrorick) [reply nav2_theta_star_planner/src/theta_star.cpp:227]: I believe this follows the same line of thinking of the comment on line [L214](https://github.com/ros-navigation/navigation2/pull/5908/changes/BASE..97bf9a918b7da9ba36924cf097d9abe893958371#diff-d6da263ed0c843bc5c6fbf8f30598edbd177d0f2a5cbee11ab660b4c1ba05f2bL214) for the `dy == 0` case.

File: nav2_theta_star_planner/src/theta_star.cpp (hunk 3 above)
Comment (SteveMacenski) [reply nav2_theta_star_planner/src/theta_star.cpp]: @astrorick I believe this is still true

**Comment by SteveMacenski** [comment]
My comment above covers 2x open points on redundant start/end checks & an edge case that may not be handled correctly: https://github.com/ros-navigation/navigation2/pull/5908#issuecomment-4202884149

Otherwise LGTM and if you rebase, we can merge this in this week hopefully! Thanks for your time and iteration cycles :-)

**Comment by astrorick** [comment]
@SteveMacenski

I believe you are right on the redundancy of the `isSafe` call for both `(x0, y0)` and `(x1, y1)` cells. However `resetParent` resets the value of `los_cost` and computes a new `g_cost` from scratch every time. If I interpreted this correctly, the only way for the cost of `(x0, y0)` to end up in `g_cost` is by checking the cell in `losCheck` at the start (when `cx = x0` and `cy = y0`), thus making that call necessary. By looking at the original implementation of this function (or the one you proposed earlier) I realized that the check on the target cell `(x1, y1)` was never performed. The safety of that cell should be already guaranteed by `setNeighbors` so my proposal is to keep the start and drop the end instead. Alternatively, we could keep the end check for added redundancy but use the API that does not accumulate `sl_cost` when calling `isSafe` there, if that is a concern.

Regarding the second point, the new implementation uses the _strict_ Bresenham variant, which differs from the _non-strict_ variant in how the algorithm handles lines that pass exactly through grid corners between their endpoints (for example, `(0, 0)` -> `(10, 5)`). The order of horizontal/vertical/diagonal movements only changes for this type of edge cases between the two. I believe the only difference if we use the non-strict variant is slightly more conservative paths, but both variants should work as expected. I did some tests locally and I could not find significant changes. I'll add this to the next commit.

Let me know what you think, I should be able to commit the missing pieces by the end of the week if everything looks good to you!

Thanks for the patience.

**Comment by SteveMacenski** [comment]
> my proposal is to keep the start and drop the end instead

A-OK please do :-) 

> I'll add this to the next commit.

OK

Please add these changes, test again, and I would aim to merge this this week so it can be included in the Lyrical release :-) Please rebase your PR though so CI passes, we changed a bunch and CI isn't going to work if you don't bring in recent changes.

**Review by SteveMacenski — APPROVED** [review]
I'm rerunning CI from what I assume failed as a fluke - can you confirm this passes your test /  benchmark cases now? Can merge then :-) 

Thanks so much for iterating over these months, it is appreciated!

**Comment by astrorick** [comment]
I can confirm all tests are go on my machine even after the latest changes.

Thanks for the patience and the support you provided over the course of these months!
